micro-rdk/src/common/data_manager.rs

```diff
@@ -262,43 +252,128 @@ where
+                        // feels too specific, so we'll return to exit the function early
+                        break;
+                    }
+                };
+                // if the first message is empty, we've reached the end of the store region
+                // and it's time to move on to the next collector
+                if next_message.is_empty() {
+                    break;
+                } else if next_message.len() > MAX_SENSOR_CONTENTS_SIZE {
                     log::error!(
-                        "error decoding readings for collector key ({:?}): {:?}",
-                        collector_key,
-                        err
+                        "message encountered that was too large (>32K bytes) for collector {:?}",
+                        collector_key
                     );
-                    continue;
+                } else {
+                    current_chunk.push(next_message);
                 }
-            };
-            if !data.is_empty() {
-                let upload_request = DataCaptureUploadRequest {
-                    metadata: Some(UploadMetadata {
-                        part_id: self.part_id.clone(),
-                        component_type: collector_key.component_type.clone(),
-                        r#type: DataType::TabularSensor.into(),
-                        component_name: collector_key.r_name.clone(),
-                        method_name: collector_key.method.to_string(),
-                        ..Default::default()
-                    }),
-                    sensor_contents: data,
+                // we want to fill current_chunk until its size reaches just under
+                // MAX_SENSOR_CONTENTS_SIZE
+                let upload_success = loop {
+                    let next_message = match reader.read_next_message() {
+                        Ok(msg) => msg,
+                        Err(err) => {
+                            log::error!(
+                                "error reading message from store for collector key ({:?}): {:?}",
+                                collector_key,
+                                err
+                            );
+                            // we don't want to panic, and creating an AppClientError variant for this case
+                            // feels too specific, so we'll return to exit the function early
+                            break false;
+                        }
+                    };
+                    if next_message.len() > MAX_SENSOR_CONTENTS_SIZE {
+                        log::error!(
+                            "message encountered that was too large (>32K bytes) for collector {:?}",
+                            collector_key
+                        );
+                        continue;
+                    }
+                    let current_chunk_size: usize = current_chunk.iter().map(|m| m.len()).sum();
```

Comment:
I think this is N^2 in number of messages that end up in the chunk since you will re-compute the sum after each iteration. It is definitely safer and I hesitate to add more complexity to this already fairly complex method, but it does seem like you could avoid the N^2 by tracking the current total.